### PR TITLE
Parse gcc extensions

### PR DESCRIPTION
**Pull request by ametselaar**
This branch adds support for parsing various gnu extensions and fixes the problem with #define B(N) 0x ## N.
The handling of alternate keywords such as __asm may not be optimal. 
Can you please review these change and give comments if changes are needed.

### CODE REVIEW

**Comment by wenns** [comment]
Thanks for the PR, Im really looking forward to your changes, especially the 'resolving ambiguity by looking ahead'-peace I spotted in here. However, we're in the process of getting 0.9.2 out of the door, and this PR contains complex and big changes. I would like to come back to you and look at them one by one after the release.

**Comment by guwirth** [comment]
@ametselaar thx. Think also this is a very useful pull request with many good solutions.

Some hints from my side:
* in the past we preferred prepossessor solutions over grammar extension see https://github.com/wenns/sonar-cxx/wiki/Dealing-with-compiler-specific-code-peaces and https://github.com/wenns/sonar-cxx/blob/master/sonar-cxx-plugin/src/main/resources/macros/VS10Macros.h to keep the grammar easy
* This is one big pull request with many solutions. Would be easier to split it to specific solutions. It's easier to test, merge and document.

**Comment by guwirth** [comment]
BTW @ametselaar you seems to be very experienced with SSLR, maybe you have an idea how to solve our biggest parser issue #49?

**Comment by ametselaar** [comment]
Postponing the parser changes to version 0.9.3 is okay.
I have some other changes which may be integrated earlier;
- a parameter count metric for which I had to reintroduce a Method Resource
and
- metrics that count various types of valgrind measures, that make it
easier to distinguish leaks from problems like freeing a pointer twice

I have also looked into parsing things like `vector<vector<int>>' (issue
#49), and I have a version that can cope with it, but the AST it produces
is not equal to the AST for `vector<vector<int> >'. It builds upon my other
parser improvements and will probably need to wait for 0.9.3 as well.

2015-01-04 10:59 GMT+01:00 Günter Wirth <notifications@github.com>:

**Comment by jmecosta** [comment]
i would be interested to try the fix for 0.9.3 i can run on our code base to detect any regressions

File: cxx-squid/src/main/java/org/sonar/cxx/api/CxxKeyword.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
   ALIGNAS("alignas"),
   ALIGNOF("alignof"),
   ASM("asm"),
+  __ATTRIBUTE__("__attribute__"), // gcc extension
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/api/CxxKeyword.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 2]
```diff
@@ -461,6 +464,7 @@ private static void expressions(LexerfulGrammarBuilder b) {
       );
 
     b.rule(unaryExpression).is(
+      b.optional("__extension__"),  // marker for gcc extensions
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 3]
```diff
@@ -553,6 +560,7 @@ private static void expressions(LexerfulGrammarBuilder b) {
       ).skipIfOneChild();
 
     b.rule(assignmentExpression).is(
+        b.optional("__extension__"), // marks gcc extensions
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 4]
```diff
@@ -741,8 +750,8 @@ private static void declarations(LexerfulGrammarBuilder b) {
       b.firstOf(CxxKeyword.REGISTER, CxxKeyword.STATIC, CxxKeyword.THREAD_LOCAL, CxxKeyword.EXTERN, CxxKeyword.MUTABLE)
       );
 
-    b.rule(functionSpecifier).is(
-      b.firstOf(CxxKeyword.INLINE, CxxKeyword.VIRTUAL, CxxKeyword.EXPLICIT)
+    b.rule(functionSpecifier).is( // extern inline and alternate keywords for inline are gcc extensions
+      b.firstOf(CxxKeyword.INLINE, "__inline__", "__inline", CxxKeyword.VIRTUAL, CxxKeyword.EXPLICIT)
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 5]
```diff
@@ -771,6 +780,8 @@ private static void declarations(LexerfulGrammarBuilder b) {
     b.rule(simpleTypeSpecifier).is(
       b.firstOf(
         "char", "char16_t", "char32_t", "wchar_t", "bool", "short", "int", "long", "signed", "unsigned", "float", "double", "void", "auto",
+        b.sequence(b.firstOf("typeof", "__typeof", "__typeof__"), "(",
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 6]
```diff
@@ -879,9 +890,18 @@ private static void declarations(LexerfulGrammarBuilder b) {
       b.firstOf(
         b.sequence(CxxKeyword.ASM, "(", STRING, ")", ";"),
         b.sequence(CxxKeyword.ASM, "{", b.oneOrMore(b.nextNot(b.firstOf("}", EOF)), b.anyToken()), "}", b.optional(";")),
-        b.sequence(CxxKeyword.ASM, b.oneOrMore(b.nextNot(b.firstOf(";", EOF)), b.anyToken()), ";")
+        b.sequence(CxxKeyword.ASM, b.oneOrMore(b.nextNot(b.firstOf(";", EOF)), b.anyToken()), ";"),
+        b.sequence(b.firstOf(CxxKeyword.ASM, "__asm__", "__asm"),
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 7]
```diff
@@ -879,9 +890,18 @@ private static void declarations(LexerfulGrammarBuilder b) {
...
+            b.optional(b.firstOf(CxxKeyword.VOLATILE, "__volatile__", "__volatile")),
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 8]
```diff
@@ -889,6 +909,8 @@ private static void declarations(LexerfulGrammarBuilder b) {
     b.rule(attributeSpecifier).is(
         b.firstOf(
             b.sequence("[", "[", attributeList, "]", "]"),
+            b.sequence(b.firstOf(CxxKeyword.ASM,  "__asm__", "__asm"), "(", STRING, ")"), // gcc extension
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 9]
```diff
@@ -961,28 +984,32 @@ private static void declarators(LexerfulGrammarBuilder b) {
...
+            b.sequence("&&", b.zeroOrMore(b.firstOf(attributeSpecifierSeq, restrictQualifier))),
+            b.sequence(nestedNameSpecifier, "*", b.zeroOrMore(b.firstOf(attributeSpecifierSeq, cvQualifierSeq, restrictQualifier)))
         )
         );
 
     b.rule(cvQualifierSeq).is(b.oneOrMore(cvQualifier));
 
     b.rule(cvQualifier).is(
-        b.firstOf(CxxKeyword.CONST, CxxKeyword.VOLATILE)
+        b.firstOf(CxxKeyword.CONST, CxxKeyword.VOLATILE, "__const") // some gcc headers contain __const
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

File: cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java [hunk 10]
```diff
@@ -961,28 +984,32 @@ private static void declarators(LexerfulGrammarBuilder b) {
...
         );
 
     b.rule(refQualifier).is(
         b.firstOf("&", "&&")
         );
+    
+    b.rule(restrictQualifier).is(
+        b.firstOf("restrict", "__restrict__", "__restrict")
```
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java]: Prefer preprocessor solution: https://github.com/wenns/sonar-cxx/wiki/Misc.-notes-on-development

**Comment by guwirth** [comment]
Outdated

**Pull request closed without merging**